cpp/src/arrow/result.h

```diff
@@ -223,19 +257,30 @@ class Result : public util::EqualityComparable<Result<T>> {
   ///
   /// \param other The Result object to assign from and set to a non-OK
   /// status.
-  Result& operator=(Result&& other) {
+  Result& operator=(Result&& other) noexcept {
     // Check for self-assignment.
     if (this == &other) {
       return *this;
     }
-    AssignVariant(std::move(other.variant_));
-    other.variant_ = "Value was moved to another Result.";
-
+    Destroy();
```

Comment:
Ditto with `std::swap`, you wouldn't need any of the logic. Destructor of `other` would be properly cleanup upon going out of scope.